spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/SparkTable.java

```diff
@@ -248,21 +268,25 @@ public ScanBuilder newScanBuilder(CaseInsensitiveStringMap options) {
       icebergTable.refresh();
     }
 
-    CaseInsensitiveStringMap scanOptions = addSnapshotId(options, snapshotId);
-    return new SparkScanBuilder(sparkSession(), icebergTable, snapshotSchema(), scanOptions);
+    CaseInsensitiveStringMap scanOptions =
+        branch != null ? options : addSnapshotId(options, snapshotId);
+    return new SparkScanBuilder(
+        sparkSession(), icebergTable, branch, snapshotSchema(), scanOptions);
   }
 
   @Override
   public WriteBuilder newWriteBuilder(LogicalWriteInfo info) {
     Preconditions.checkArgument(
-        snapshotId == null, "Cannot write to table at a specific snapshot: %s", snapshotId);
+        branch != null || snapshotId == null,
```

Comment:
Would branch and snapshot ID ever be set at the same time? Those are different constructors so I think we know they are mutually exclusive.

If so, then this can be the original check that `snapshotId` must be null.

Reply:
updated!